frontend/app/components/work-packages/wp-attachments-formattable-field/wp-attachments-formattable.directive.ts

```diff
@@ -0,0 +1,131 @@
+import {WpAttachmentsService} from './../wp-attachments/wp-attachments.service';
+import {WorkPackageResource} from './../../api/api-v3/hal-resources/work-package-resource.service'
+import {InsertMode, ViewMode} from './wp-attachments-formattable.enums'
+import {DropModel, EditorModel, MarkupModel, FieldModel, SingleAttachmentModel} from './wp-attachments-formattable.models'
+
+export class WpAttachmentsFormattableController {
+    private viewMode: ViewMode = ViewMode.SHOW;
+
+    constructor(protected $scope: ng.IScope,
+                protected $element: ng.IAugmentedJQuery,
+                protected $rootScope: ng.IRootScopeService,
+                protected $location: ng.ILocationService,
+                protected wpAttachments: WpAttachmentsService,
+                protected $timeout: ng.ITimeoutService) {
+
+      $element.get(0).addEventListener('drop', this.handleDrop);
+      $element.bind('dragenter', this.prevDefault)
+        .bind('dragleave', this.prevDefault)
+        .bind('dragover', this.prevDefault);
+
+    }
+
+    public handleDrop = (evt: DragEvent): void => {
+        evt.preventDefault();
+        evt.stopPropagation();
+
+        const textarea: ng.IAugmentedJQuery = this.$element.find('textarea');
+        this.viewMode  = (textarea.length > 0) ? ViewMode.EDIT : ViewMode.SHOW;
+
+        const workPackage: WorkPackageResource = this.$scope.workPackage;
+        const dropData: DropModel = new DropModel(this.$location, evt.dataTransfer, workPackage);
+      
+        var description: any;
+
+        if (this.viewMode === ViewMode.EDIT) {
+            description = new EditorModel(textarea, new MarkupModel());
+        }
+        else {
+            description = new FieldModel(workPackage, new MarkupModel());
+        }
+
+        if (angular.isUndefined(dropData.webLinkUrl) && angular.isUndefined(dropData.files))
+            return;
+
+        if(dropData.isUpload){
```

Comment:
Please extract the following conditions into separate helper methods. The condition being that long results in additional complexity for others to understand and modify that code, should it ever become necessary.

Looking at `handleDrop`, I should be able to glance over the code and see the several cases and paths that are handled there. The actual uploading, sorting, error handling, etc., should be refactored into methods.